src/geometry/manifold/basic_smooth_bundle.lean

```diff
@@ -0,0 +1,570 @@
+             (range I.to_fun) (I.to_fun x),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      simpa using hx },
+    have D : fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ j.1.inv_fun ∘ I.inv_fun)
+               (I.inv_fun ⁻¹' (j.1.symm.trans u.1).source ∩ range I.to_fun)
+               ((I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun) (I.to_fun x)) =
+             fderiv_within 𝕜 (I.to_fun ∘ u.1.to_fun ∘ j.1.inv_fun ∘ I.inv_fun)
+               (range I.to_fun)
+               ((I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun) (I.to_fun x)),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      rw [local_homeomorph.trans_source] at hx,
+      simp [mem_inter_iff, -mem_image, -mem_range, mem_range_self],
+      exact hx.2 },
+    have E : fderiv_within 𝕜 (I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+               (I.inv_fun ⁻¹' ((i.1.symm.trans j.1).trans (j.1.symm.trans u.1)).source ∩ range I.to_fun)
+               (I.to_fun x) =
+             (fderiv_within 𝕜 (I.to_fun ∘ j.1.to_fun ∘ i.1.inv_fun ∘ I.inv_fun)
+               (range I.to_fun)
+               (I.to_fun x)),
+    { rw inter_comm,
+      apply fderiv_within_inter _ (I.unique_diff _ (mem_range_self _)),
+      apply I.continuous_inv_fun.continuous_at.preimage_mem_nhds
+        (mem_nhds_sets (local_homeomorph.open_source _) _),
+      simpa using hx },
+    rw [B, C, D, E] at A,
+    rw A,
+    assume v,
+    simp,
+    unfold_coes,
+    simp
+  end }
+
+/-- The tangent bundle to a smooth manifold, as a plain type. -/
+def tangent_bundle := (tangent_bundle_core I M).to_topological_fiber_bundle_core.total_space
+
+/-- The projection from the tangent bundle of a smooth manifold to the manifold. As the tangent
+bundle is represented internally as a product type, the notation `p.1` also works for the projection
+of the point `p`. -/
+def tangent_bundle.proj : tangent_bundle I M → M :=
+(tangent_bundle_core I M).to_topological_fiber_bundle_core.proj
+
+variable {M}
+
+/-- The tangent space at a point of the manifold M. It is just E. -/
+def tangent_space (x : M) : Type* :=
+(tangent_bundle_core I M).to_topological_fiber_bundle_core.fiber x
+
+section tangent_bundle_instances
+
+/- In general, the definition of tangent_bundle and tangent_space are not reducible, so that type
+class inference does not pick wrong instances. In this section, we record the right instances for
+them, noting in particular that the tangent bundle is a smooth manifold. -/
+variable (M)
+local attribute [reducible] tangent_bundle
+
+instance : topological_space (tangent_bundle I M) := by apply_instance
```

Comment:
Can you avoid some boilerplate with `@[derive [topological_space, manifold (H × E), smooth_manifold_with_corners I.tangent]]` on the definition of `tangent_bundle`? Or does that not work here? Same for the instances below, although you might have to move the reducibility attribute.

Reply:
The derive works fine, but it breaks something later on. With the current version, if I have `(p : tangent_bundle I H)` then `simp [chart_at]` transforms `chart_at (H × E) p` to its definition `(basic_smooth_bundle_core.chart (tangent_bundle_core I H)`. If instead I use `@derive`, then `simp [chart_at]` does nothing, and neither does `dsimp [chart_at]`. No idea why. So for now I have not made the change.